### DISCUSSION TITLE
Having issues with button background changing shape / dimensions when disabling.

### DISCUSSION THREAD

**Discussion by sethstenzel** [TomSchimansky/CustomTkinter#1838] [Q&A] [answered]
When disabling a button, the background of the button is sometimes missing. The background element of the button is changing dimensions, but there is nothing in the code that changes the shape or size of buttons in the application.

Any help resolving the issue would be greatly appreciated.

**Button code:**
```
    # ~ ##############################################################
    # & Connect Serial Button
    connect_serial_button = customtkinter.CTkButton(
        master=tab,
        text="CONNECT",
        command=partial(run_threaded_task, connect_serial_monitor),
        width=50,
        height=30,
        corner_radius=0,
        fg_color=ORANGE,
        text_color=DARKBLUE,
        hover_color=LIGHTORANGE,
        font=("Segoe UI bold", 14),
    )
    connect_serial_button.place(x=15, y=95)
    app.globals["serial_monitor_tab"]["connect_serial_button"] = connect_serial_button
 
```

**Button disable code for debugging:**
```
    i = 20
    while i > 0:
        i -= 1
        app.globals["serial_monitor_tab"]["connect_serial_button"].configure(state="disabled")
        print("disabled")
        time.sleep(0.5)

        app.globals["serial_monitor_tab"]["connect_serial_button"].configure(state="normal")
        print("enabled")
        time.sleep(0.5)
```

And the results of the above code:
![image](https://drive.google.com/uc?export=view&id=1-_bG1N29qshPfhW_oFAp_V7Sw5ZKdK-1)

**Comment by sethstenzel** [comment]
As a work around I appear to be able to negate the issue by repassing the width and height each time I enable or disable, and I do have to pass the dimension when both enabling and disabling or the size issue can occur.

This is very tedious as I sometimes disable every element on a tab.

**WORK AROUND**
```
    i = 20
    connect_serial_button = app.globals["serial_monitor_tab"]["connect_serial_button"]
    while i > 0:
        i -= 1
        connect_serial_button.configure(
            state="disabled",
            width=connect_serial_button.cget("width"),
            height=connect_serial_button.cget("height"),
        )
        print("disabled")
        time.sleep(0.5)

        connect_serial_button.configure(
            state="normal",
            width=connect_serial_button.cget("width"),
            height=connect_serial_button.cget("height"),
        )
        print("enabled")
        time.sleep(0.5)
```

**Comment by ElectricCandlelight** [comment]
'''command=partial(run_threaded_task, connect_serial_monitor),'''

Are you running code in another thread? Tkinter operations should all be performed on the same thread. Tkinter isn't thread safe.

**Reply by sethstenzel** [reply]
Hmm, I am doing it in another thread. I guess I just dreamed up the idea that using a lock would resolve any thread-related issues if a thread changed an element.

**Comment by sethstenzel** [comment, marked as answer]
It may very well be a threading issue, I don't know. However, the issue appears to be related to the button background automatic resizing for text that is longer than the original button dimensions. So my buttons were sized too small for the text on them. They auto-sized for the text correctly, but somewhere between enabling them and disabling them on a separate thread, it seems to fail to resize the background correctly sometimes.

**Simply making sure they were the correct width and height for the intended text appears to have resolved the issue.**
